demos/yielding/README.md

```diff
@@ -0,0 +1,112 @@
+* `do_scheduler` (in `src/kernel/thread.c`)
+* `yield` (in `src/user/thread.c`)
+
+## `add_thread_from_worker`
+
+In AMT, we try to maintain all of the threads in an array called `all_threads`.
+
+By calling this function, the AMT will create a new thread instance with the given arguments, initialize the permissions, and put the function in. 
+
+After the creation of the thread, the entity will be put into `all_threads` array and wait for the utilization of the CPU.
+
+## `do_scheduler`
+
+The job of the scheduler is to choose the next thread to make use of the CPU. 
+
+In AMT, the scheduler simply walks through the `all_threads` array to find the next thread.
+
+To run the thread chosen by the scheduler and restore the function call stack of the current thread, we need to save the information of this thread first.
+
+To achieve the goal, the context switch routine written in assembly (`yield.S`) will be executed to find the next task. 
+
+And the task of the thread to be switched will be suspended until the thread has been chosen to run
+
+## `yield`
+
+To yield the utilization, the thread calls the syscall yield and set the kind of yielding as `YIELD_ANY`, which means that we do not specify the thread to acquire the CPU. As a result, the scheduler will help us to choose the next thread to run.
+
+And the current thread will have to wait until next time a thread had yielded the utilization to it.
+
+## The progress of yielding demo
+
+At the beginning of the demo, to observe the behaviour of the scheduler, we use `set_kernel_config` to allow the scheduler to print out the log. 
+
+After setting up the config, we use `add_thread_from_worker` to create a new thread running the `thread_worker`. 
+
+When the initialization is complete, the first thread starts to run into an endless loop and repeatedly yield to others.
+
+Since there are only two threads, the scheduler will directly choose the thread running `thread_worker` to execute.
+
+In `thread_worker` function
+
+```cpp
+__attribute__((noreturn)) void thread_worker_1() {
+  while (1) {
+    for (int i = 0;; ++i) {
+      if (i == 2) {
+        log_event("working");
+        log_event("exiting");
+        exit(0);
+      }
+      yield();
+    }
+  }
+}
+```
+
+We know that the thread will try to call `yield` twice and call `exit` to terminate the demo.
+
+To explain the log message of "yielding" and "resuming", we can check the implementation of `yield`.
+
+In file `scr/user/thread.c`,
+
+```cpp
+void yield(void) {
+  log_event("yielding");
+  DO_SYSCALL_2(yield, INVALID_THREAD, YIELD_ANY);
+  log_event("resuming");
+}
+```
+
+We can find that before yielding, the thread will dump the log message "yielding", and when the utilization is back to this thread, the log message "resuming" will be dumped on the screen and the thread will resume running the given subroutine.
+
+According to the behaviour of the two threads, the result of this demo will be:
```

Comment:
Similar to the source chunks I don't think including log output is going to be useful. Assuming the reader has the repo open with the docs, they can read the expected output themselves.

I would prefer a higher level description of the demo. Here you could explain where the "yielding", "resuming" messages come from (which I think is a great idea for the first demo people will read) then give a description of the rest.

Like (super rough wording): "thread 0 repeatedly yields to thread 1. Thread 1 counts the number of times it has run. If it has run twice then it calls exit() to end the demo. This shows that threads are resumed from the point they yield from, if this were not the case then the demo would effectively become an infinite loop"

Maybe "run" is too imprecise, it can mean it was restarted (another imprecise word). So maybe "number of times it has been yielded to", which makes it clear that it doesn't restart. (or "has has had CPU time" whatever makes sense to you)

Side note: The docs for Linux PTRACE_CONT use the word "restart" which confused me a good deal, that's why I bring that up.